growthbook/growthbook
# Type-safe features in the SDK & CLI docs

### Features and Changes

Adds support for `GrowthBook` to take an optional generic type argument to define the features.

Implements the following methods:

- `getFeatureValue(knownKey, defaultValue)`
- `evalFeature(knownKey)`
- `isOn(knownKey)`
- `isOff(knownKey)`

You create a `GrowthBook` instance as usual and you pass the type definitions in as the generic argument parameter.

If no generic argument is passed, the methods will continue to have the current behaviour (inferring type from passed in argument).


### Dependencies

- [x] https://github.com/growthbook/growthbook/pull/1022 – there's a link in these docs that link to the CLI page. To ensure the CLI link is not broken, I merged the docs for the CLI into this PR


### Testing

See tests

- See CLI docs: https://docs-6choy8e04-growthbook.vercel.app/tools/cli
- See JS SDK - TypeScript docs: https://docs-6choy8e04-growthbook.vercel.app/lib/js#typescript

### Screenshots

In the below screenshots, you can see the editor hints, as well as failed test runs when incorrect keys are passed.

<img width="927" alt="Screen Shot 2023-02-23 at 2 55 40 PM" src="https://user-images.githubusercontent.com/113377031/221049787-bb581cb4-ae29-49f4-938f-0c0adb5dcc07.png">

<img width="1039" alt="Screen Shot 2023-02-23 at 2 57 15 PM" src="https://user-images.githubusercontent.com/113377031/221049799-0222cd62-0bf7-4589-82f0-a70728f3d67f.png">



<img width="2411" alt="image" src="https://user-images.githubusercontent.com/113377031/221049654-7376e26d-f4c0-412b-9dda-25ca45f7fb41.png">
